pkg/gatewayserver/io/ws/ws.go

```diff
@@ -63,17 +54,17 @@ type srv struct {
 	server               io.Server
 	webServer            *echo.Echo
 	upgrader             *websocket.Upgrader
-	tokens               io.DownlinkTokens
 	useTrafficTLSAddress bool
 	wsPingInterval       time.Duration
 	cfg                  Config
+	formatter            Formatter
 }
 
 func (*srv) Protocol() string            { return "basicstation" }
 func (*srv) SupportsDownlinkClaim() bool { return false }
 
-// New creates the Basic Station front end.
-func New(ctx context.Context, server io.Server, cfg Config) *echo.Echo {
+// New creates the LoRa Basics Station front end.
+func New(ctx context.Context, server io.Server, formatter Formatter, cfg Config) *echo.Echo {
```

Comment:
OK so `Formatter` implementation is a singleton.

I think that the session should be passed via context to the formatter, and not make the formatter implementation manage session lifetime via connect/disconnect.